File: src/dynhung.cpp
#include "dynhung.h"
#include <iostream> //cout
#include <algorithm> //transform, fill_n

assignment_problem::assignment_problem(std::vector<double> cost_matrix) {
	C = cost_matrix;
	n = (int) sqrt(C.size());
	ITER = 0;

	if (n != sqrt(C.size())) {
		throw std::length_error("cost matrix not square");
	}

	hungarian();
}


void assignment_problem::hungarian() {
	//Fill with zeros
	covered_rows.resize(n);
	covered_cols.resize(n);

	//Initialise assignments
	std::vector<int> unassigned(n, UNASSIGNED);
	row_assign = unassigned;
	col_assign = unassigned;

	//Copy the cost matrix into the slack matrix
	c = C;
	
	//Start at the beginning
	step = 1;
	assignments = 0;
	
	while (assignments < n) {
		if (step == 1) {
			step_one();
		} else if (step == 2) {
			step_two();
		} else if (step == 3) {
			step_three();
		} else if (step == 4) {
			step_four();
		}
				
		ITER++;
	}

	compute_cost();
}

void assignment_problem::update_hungarian() {
	//Recompute the slack matrix
	c.clear();
	c.resize(n*n);

	//Repair the slack matrix (fix tiny errors due to rounding errors)
	for (int i = 0; i < n; i++) {
		for (int j = 0; j < n; j++) {
			double slack = C[i*n + j] - row_dual[i] - col_dual[j];
			if (row_assign[i] == j) {
				slack = 0;
			}
			if (slack < 0) {
				slack = 0;
			}
			if (slack < 1e-15) {
				slack = 0;
			}
			
			c[i*n + j] = slack;
		}
		std::cout << std::endl;
	}	

	//Continue with the Hungarian
	step = 3;
	ITER = 0;

	while (assignments < n) {
		if (step == 3) {
			step_three();
		} else if (step == 4) {
			step_four();
		}
		ITER++;
	}

	compute_cost();
}

void assignment_problem::compute_cost() {
	cost = 0;
	for (int i = 0; i < n; i++) {
		int j = row_assign[i];
		cost += C[i*n + j];
	}
}

//Note: the user must indicate which rows changed!
void assignment_problem::update_rows(std::vector<double> cost_matrix, std::vector<int> rows_changed) {
	for (unsigned i = 0; i < rows_changed.size(); i++) {
		if (rows_changed[i] < 0 || rows_changed[i] >= n) {
			throw std::domain_error("Row index is out of bounds.");
		}		
	}
	C = cost_matrix;
	
	for (unsigned i = 0; i < rows_changed.size(); i++) {
		int row_change = rows_changed[i];
				
		//Remove the ith edge
		int the_col = row_assign[row_change];
		if (the_col != UNASSIGNED) {
			assignments--;
			col_assign[the_col] = UNASSIGNED;
			row_assign[row_change] = UNASSIGNED;
		}
				

		//Update the ith dual value
		double smallest = DBL_MAX;
		for (int j = 0; j < n; j++) {
			double slack = C[row_change*n + j] - col_dual[j];
			if (slack < smallest) {
				smallest = slack;
			}
		}
		row_dual[row_change] = smallest;
	}
	
	update_hungarian();
}

//Note: the user must indicate which columns changed!
void assignment_problem::update_cols(std::vector<double> cost_matrix, std::vector<int> cols_changed) {
	for (unsigned i = 0; i < cols_changed.size(); i++) {
		if (cols_changed[i] < 0 || cols_changed[i] >= n) {
			throw std::domain_error("Column index is out of bounds.");
		}		
	}
	C = cost_matrix;
	
	for (unsigned i = 0; i < cols_changed.size(); i++) {
		int col_change = cols_changed[i];
				
		//Remove the ith edge
		int the_row = col_assign[col_change];
		if (the_row != UNASSIGNED) {
			assignments--;
			row_assign[the_row] = UNASSIGNED;
			col_assign[col_change] = UNASSIGNED;
		}

		//Update the ith dual value
		double smallest = DBL_MAX;
		for (int k = 0; k < n; k++) {
			double slack = C[k*n + col_change] - row_dual[k];
			if (slack < smallest) {
				smallest = slack;
			}
		}
		col_dual[col_change] = smallest;
	}
	
	update_hungarian();
}

void assignment_problem::step_one() {
	col_dual.clear();
	col_dual.resize(0);
	row_dual.clear();
	row_dual.resize(0);

	//Allocate memory
	col_dual.reserve(n);
	row_dual.reserve(n);

	//Subtract smallest element from each row
	for (int i = 0; i < n; i++) {
		double smallest = DBL_MAX;
		for (int j = 0; j < n; j++) {
			if (c[i*n + j] < smallest) {
				smallest = c[i*n + j];
			}
		}
		
		//Record the dual value
		row_dual.push_back(smallest);

		for (int j = 0; j < n; j++) {
			c[i*n + j] -= row_dual[i];
		}
	}
	
	//Subtract smallest element from each column
	for (int j = 0; j < n; j++) {
		double smallest = DBL_MAX;
		for (int i = 0; i < n; i++) {
			if (c[i*n + j] < smallest) {
				smallest = c[i*n + j];
			}
		}
	

		//Record the dual value
		col_dual.push_back(smallest);

		for (int i = 0; i < n; i++) {
			c[i*n + j] -= col_dual[j];
		}
	}

	step = 2;
}


void assignment_problem::step_two() {
	//Count the number of zeros per row and column.
	std::vector<int> row_zeros(n);
	std::vector<int> col_zeros(n);
	
	for (int i = 0; i < n; i++){
		for (int j = 0; j < n; j++) {
			if (c[i*n + j] == 0) {
				row_zeros[i]++;
				col_zeros[j]++;
			}
		}
	}
	
	//Find the unassigned row with zeros that has the least number of zeros
	bool done = false;
	while (!done) {
		int smallest = INT_MAX;
		int row = NONE;
		for (int i = 0; i < n; i++) {
			if (row_assign[i] == UNASSIGNED && row_zeros[i] > 0 && row_zeros[i] < smallest) {
				row = i;
				smallest = row_zeros[i];
			}
		}

		if (row == NONE) {
			done = true;
		} else {
			//Find the unassigned column that has a zero in row r with the least number of zeros
			int col = NONE;
			smallest = INT_MAX;
			for (int j = 0; j < n; j++) {
				if (c[row*n + j] == 0 && col_zeros[j] < smallest && col_assign[j] == UNASSIGNED) {
					col = j;
					smallest = col_zeros[j];
				}
			}

			if (col != NONE) {
				//Assign this row to this column
				assignments++;
				row_assign[row] = col;
				col_assign[col] = row;

				//Remove this zero from the counters
				col_zeros[col] = 0;
				for (int i = 0; i < n; i++) {
					if (c[i*n + col] == 0) {
						row_zeros[i]--;
					}
				}
			}

		}

	}
	
	step = 3;
}


void assignment_problem::step_three() {
	//Clear all column covers
	covered_cols.clear();
	covered_cols.resize(n);
	covered_rows.clear();
	covered_rows.resize(n);
	
	//Initialise row marks
	std::vector<int> marked_rows;
	marked_rows.resize(n);
	
	//Reset all row covers
	for (int i = 0; i < n; i++) {
		if (row_assign[i] == UNASSIGNED) {
			covered_rows[i] = UNCOVERED;
			marked_rows[i] = MARKED;
		} else {
			covered_rows[i] = COVERED;
		}
	}

	bool done = false;
	while (!done) {
		//Find marked rows
		bool found = false;
		int i = 0;
		while (!found && i < n) {
			if (marked_rows[i] == MARKED) {
				found = true;
			} else {
				i++;
			}
		}

		if (!found) {
			done = true;
			step = 4;
		} else {			
			//Find uncovered zeros in this row
			for (int j = 0; j < n; j++) {
				if (c[i*n + j] == 0 && covered_cols[j] == UNCOVERED) {
					if (col_assign[j] != UNASSIGNED) {
						//This column is currently assigned to
						int assigned_row = col_assign[j];
						//Uncover and mark that row and cover this column
						covered_rows[assigned_row] = UNCOVERED;
						marked_rows[assigned_row] = MARKED;
						covered_cols[j] = COVERED;
					} else {
						//We are going to assign this row to this column
						if (row_assign[i] == UNASSIGNED) {
							assignments++;
						} else {
							//This row is currently assigned to 
							int assigned_col = row_assign[i];
							//But no longer, so clear the reverse assignment
							col_assign[assigned_col] = UNASSIGNED;
						}

						//Assign this row to this column
						row_assign[i] = j;
						col_assign[j] = i;

						done = true;
						step = 3;
					}
				}
			}
			
			marked_rows[i] = UNMARKED;

		}
	}
}


void assignment_problem::step_four() {
	//Find the smallest uncovered value (TODO: in O(n) rather than O(n^2))
	//Subtract from each element of each uncovered column.
	//Add to each element of each covered row. Go to step 4.
	double smallest = find_smallest_uncovered_val();
	for (int i = 0; i < n; i++) {
		for (int j = 0; j < n; j++) {
			if (covered_rows[i] == COVERED) {
				c[i*n + j] += smallest;
			}
			if (covered_cols[j] == UNCOVERED) {
				c[i*n + j] -= smallest;
			}
		}
	}

	//Update dual values
	for (int i = 0; i < n; i++) {
		if (covered_rows[i] == UNCOVERED) {
			row_dual[i] += smallest;
		}
		if (covered_cols[i] == COVERED) {
			col_dual[i] -= smallest;
		}
	}
	
	step = 3;
}

double assignment_problem::find_smallest_uncovered_val() {
	double smallest = DBL_MAX;
	for (int i = 0; i < n; i++) {
		for (int j = 0; j < n; j++) {
			if (c[i*n + j] < smallest && covered_rows[i] == UNCOVERED && covered_cols[j] == UNCOVERED) {
				smallest = c[i*n+j];
			}
		}
	}

	
	return smallest;
}

File: R/dynhung.r
#' @useDynLib dynhung
#' @import Rcpp
NULL
loadModule("APMod", TRUE);

#' Optimal Matching
#'
#' Solve the linear sum assignment problem using the Hungarian method. 
#' @param cost_matrix a square matrix with non-negative cost values
#' @keywords assignment problem
#' @export
#' @examples #Example:
#' A <- matrix(c(1,2,3,2,4,6,3,6,9), nrow=3)
#' B <- matrix(c(1,2,3,2,4,6,3,6,0), nrow=3)
#' AP_A <- solve_assignment(A)
#' AP_B <- solve_assignment(B)
#' matching_A <- optimal_matching(AP_A)
#' matching_B <- optimal_matching(AP_B)
#' print("Optimal matching with A:")
#' print(matching_A)
#' print("Optimal matching with B:")
#' print(matching_B)
#' update_rows(AP_A, B, 3)
#' update_cols(AP_B, A, 3)
#' matching_A <- optimal_matching(AP_A)
#' matching_B <- optimal_matching(AP_B)
#' print("Optimal matching with A:")
#' print(matching_A)
#' print("Optimal matching with B:")
#' print(matching_B)
#'
#' #Compare with clue package function solve_LSAP
#' library(clue)
#' solve_LSAP(A)
#' solve_LSAP(B)
solve_assignment <- function(cost_matrix) {
	if (!is.numeric(cost_matrix)) {
		stop("Cost matrix is not numeric.")
	}
	
	#Convert matrix to 1-dimensional array
	cost_matrix <- as.vector(t(cost_matrix))
	
	n <- sqrt(length(cost_matrix))
	if (abs(n - round(n)) > .Machine$double.eps) {
		stop("Cost matrix is not square.")
	}
		
	AP <- new(assignment_problem, cost_matrix)
	# matching <- AP$get_assignment()
	
	# #Convert 0-based indexing to 1-based indexing
	# matching <- matching + 1
	
	return(AP)
}

#' Optimal Matching
#'
#' Solve the linear sum assignment problem using the Hungarian method. 
#' @param AP an object of the assignment_problem class produced with solve_assignment()
#' @keywords assignment problem
#' @export
#' @examples
#' optimal_matching()
optimal_matching <- function(AP) {
	matching <- AP$get_assignment()
	
	#Convert 0-based indexing to 1-based indexing
	matching <- matching + 1
	
	return(matching)
}

#' Optimal Cost
#'
#' Solve the linear sum assignment problem using the Hungarian method. 
#' @param AP an object of the assignment_problem class produced with solve_assignment()
#' @keywords assignment problem
#' @export
#' @examples
#' optimal_cost()
optimal_cost <- function(AP) {
	cost <- AP$get_cost()
	
	return(cost)
}

#' Update rows of the cost matrix for the assignment problem
#'
#' Solve the linear sum assignment problem using the Hungarian method. 
#' @param AP an object of the assignment_problem class produced with solve_assignment()
#' @param cost_matrix a new cost matrix
#' @param rows_changed a vector of row indices
#' @keywords assignment problem
#' @export
#' @examples
#' update_rows()
update_rows <- function(AP, cost_matrix, rows_changed) {
	if (!is.numeric(cost_matrix)) {
		stop("Cost matrix is not numeric.")
	}
	
	#Convert matrix to 1-dimensional array
	cost_matrix <- as.vector(t(cost_matrix))
	
	n <- sqrt(length(cost_matrix))
	if (abs(n - round(n)) > .Machine$double.eps) {
		stop("Cost matrix is not square.")
	}
	
	if (n != AP$get_n()) {
		stop("Cost matrix has changed in size.");
	}
	
	#Convert 1-based indexing to 0-based indexing
	rows_changed = rows_changed - 1
	
	AP$update_rows(cost_matrix, rows_changed)
}

#' Update columns of the cost matrix for the assignment problem
#'
#' Solve the linear sum assignment problem using the Hungarian method. 
#' @param AP an object of the assignment_problem class produced with solve_assignment()
#' @param cost_matrix a new cost matrix
#' @param cols_changed a vector of column indices
#' @keywords assignment problem
#' @export
#' @examples
#' update_cols()
update_cols <- function(AP, cost_matrix, cols_changed) {
	if (!is.numeric(cost_matrix)) {
		stop("Cost matrix is not numeric.")
	}
	
	#Convert matrix to 1-dimensional array
	cost_matrix <- as.vector(t(cost_matrix))
	
	n <- sqrt(length(cost_matrix))
	if (abs(n - round(n)) > .Machine$double.eps) {
		stop("Cost matrix is not square.")
	}
	
	if (n != AP$get_n()) {
		stop("Cost matrix has changed in size.");
	}
	
	#Convert 1-based indexing to 0-based indexing
	cols_changed = cols_changed - 1
	
	AP$update_cols(cost_matrix, cols_changed)
}